src/main/java/software/amazon/nio/spi/s3/S3WritableByteChannel.java

```diff
@@ -44,18 +46,31 @@ class S3WritableByteChannel implements SeekableByteChannel {
 
         try {
             var fileSystemProvider = (S3FileSystemProvider) path.getFileSystem().provider();
-            var exists = fileSystemProvider.exists(client, path);
 
-            if (exists && options.contains(StandardOpenOption.CREATE_NEW)) {
+            if (options.contains(StandardOpenOption.CREATE_NEW) && fileSystemProvider.exists(client, path)) {
                 throw new FileAlreadyExistsException("File at path:" + path + " already exists");
             }
-            if (!exists && !options.contains(StandardOpenOption.CREATE_NEW) && !options.contains(StandardOpenOption.CREATE)) {
-                throw new NoSuchFileException("File at path:" + path + " does not exist yet");
-            }
 
             tempFile = Files.createTempFile("aws-s3-nio-", ".tmp");
-            if (exists) {
-                s3TransferUtil.downloadToLocalFile(path, tempFile);
+            // this complicated download handling is the result of
+            // avoiding an existence check with a head-object request
+            if (!options.contains(StandardOpenOption.CREATE_NEW)) {
+                try {
+                    s3TransferUtil.downloadToLocalFile(path, tempFile);
+                } catch (CompletionException e) {
+                    var cause = e.getCause();
```

Comment:
You could switch on the cause being an instance of NoSuchKeyException rather than using the status code. This would make the logic more clear.

Reply:
Got it. Local stack might not be totally faithful to the API in this case. I will resolve the comment and merge. Thanks for the PR!